add: Новый пистолет "Спектр"

PiroMage opened the pull request:
<!-- Пишите **НИЖЕ** заголовков и **ВЫШЕ** комментариев, иначе ваш текст может не отобразиться. -->
<!-- В Contributing.MD вы можете найти некоторые рекомендации к оформлению пулл-реквеста. -->

## Описание
Добавление нового пистолета для СБ - Спектр
<!-- Опишите, что делает ваш Pull request. Документировать каждую деталь не требуется, просто укажите основные изменения. -->

## Причина создания ПР / Почему это хорошо для игры
https://discord.com/channels/617003227182792704/1338799136719241300/1339917472110870528
<!-- Здесь можно оставить ссылку на сообщение в #отчеты-по-предложениям, чтобы подтвердить, что предложение обсуждалось внутри Discord-сообщества. -->
<!-- Если отчёта нет, то укажите, почему это изменение положительно влияет на игру. -->
<!-- В случае исправления бага, укажите ссылку на канал в #баг-репорты-v2 или issue в репозитории. В ином случае, опишите баг и ступени для его воспроизведения. -->
<!-- Пример ссылки в Discord-сообщество : https://discord.com/channels/617003227182792704/755125334097133628/ID-сообщения -->

## Демонстрация изменений

<!-- В случае наличия изменений, влияющих на игровую часть, опишите их здесь. В случае их отсутствия, этот пункт можно удалить -->
https://discord.com/channels/617003227182792704/734823601110515882/1382512557428052120 
Ссылка на три видео (в канале тестового полигона), в которых демонстрируется функционал и работоспособность. Собственно все проверено на локалке.
## Тесты

<!-- Здесь необходимо описать шаги, которые предпринимались для тестирования изменения. Этот пункт обязателен, без него Pull request будет рассматриваться дольше. -->

PiroMage committed:
Базовый функционал

Добавление базового функционала Спектра, за основу взят код ЛРки

PiroMage commented:
Щас доделаем...

PiroMage committed:
Доработки...

Почти готово, осталось решить последние конфликты с лазерами

PiroMage commented:
Почему-то жалуется на те же самые переменные, которые уже используются в билде у других пушек... Памахити...

PiroMage committed:
Конечная...

Все работает, все проверил на локалке.

PiroMage commented:
Все, все работает... Проверено на локалке.

littleboobs commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -675,6 +675,20 @@
 
+/obj/item/ammo_box/magazine/specmaglaser
+	name = "small encased laser projector magazine"
+	desc = "Fits experimental laser ammo casings."
```
экспериментальные? Просто лазерные, как везде.
Перевод.

PiroMage replied:
Это описание таких же патронов, которые используются у ЛРки и карабина, это буквально лазер в пуле, поэтому экспериментальные, перевод будет.

littleboobs commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -30,6 +30,56 @@
+
+//Specter//
+/obj/item/gun/projectile/automatic/pistol/specter
+	name = "Specter"
+	ru_names = list(
+	NOMINATIVE = "Спектр",
+	GENITIVE = "Спектра",
+	DATIVE = "Спектру",
+	ACCUSATIVE = "Спектр",
+	INSTRUMENTAL = "Спектром",
+	PREPOSITIONAL = "Спектре"
+	)
+	desc = "Современный пистолет \"Спектр\", модернизирован для возможности стрельбы лазерными патронами. Поставляется только силовым структурам Нанотрейзен."
```
рунеймсы после описания, табуляция перед падежами

PiroMage replied:
Ну вроде поменял

littleboobs commented on code/modules/research/designs/autolathe_designs.dm:
```diff
@@ -865,6 +865,39 @@
 	build_path = /obj/item/ammo_box/c9mmte
 	category = list("hacked", "Security")
 
+/datum/design/specter/disable
+	name = "Ammo box (specter disable)"
+	id = "specter_disable"
+	build_type = AUTOLATHE
+	materials = list(MAT_METAL = 25000)
+	build_path = /obj/item/ammo_box/specter/disabler
+	category = list("initial", "Security")
+
+/datum/design/specter/laser
+	name = "Ammo box (specter laser)"
+	id = "specter_laser"
+	build_type = AUTOLATHE
+	materials = list(MAT_METAL = 35000)
+	build_path = /obj/item/ammo_box/specter/laser
+	category = list("hacked", "Security")
+
+/datum/design/spectermag_disabler
+	name = "specter magazine (disable)"
+	id = "spectermag_disabler"
+	build_type = AUTOLATHE
+	materials = list(MAT_METAL = 8000)
+	build_path = /obj/item/ammo_box/magazine/specter
+	category = list("initial", "Security")
+
+//hacked autolathe recipes
+/datum/design/spectermag_laser
+	name = "specter magazine (laser)"
+	id = "spectermag_laser"
+	build_type = AUTOLATHE
+	materials = list(MAT_METAL = 10000)
+	build_path = /obj/item/ammo_box/magazine/specter/laser
+	category = list("hacked", "Security")
+
```
рунеймсы, Ammo с маленькой буквы

PiroMage replied:
ammo box (specter disable)
ammo box (specter laser)
Переделал. Рунеймсов не будет, будет вырвиглаз в автолате (все другие вещи на англ. почему только это будет на русском?)

Antoonij replied:
у датумов нету рунеймсов + тут смысла вводить их нет, для отображения в консоли достаточно обычного name, без каких либо спряжений. Освобождаю человека от этого саггеста

littleboobs commented on code/modules/projectiles/ammunition/ammo_casings.dm:
```diff
@@ -374,6 +374,24 @@
 	muzzle_flash_range = MUZZLE_FLASH_RANGE_NORMAL
 	muzzle_flash_color = LIGHT_COLOR_DARKRED
 
+/obj/item/ammo_casing/specter/laser
+	desc = "A specter bullet casing."
+	icon_state = "laser-casing"
+	caliber = "specter"
+	materials = list(MAT_METAL = 1000)
+	projectile_type = /obj/projectile/beam/specter/laser
+	muzzle_flash_strength = MUZZLE_FLASH_STRENGTH_WEAK
+	muzzle_flash_range = MUZZLE_FLASH_RANGE_NORMAL
+	muzzle_flash_color = LIGHT_COLOR_DARKRED
+
+/obj/item/ammo_casing/specter/disable
+	desc = "A specter bullet casing."
+	icon_state = "stam-casing"
+	caliber = "specter"
+	materials = list(MAT_METAL = 800)
+	projectile_type = /obj/projectile/beam/specter/disabler
+	muzzle_flash_color = LIGHT_COLOR_LIGHTBLUE
+
```
неймы не нужны? описания перевести.

PiroMage replied:
Щас будет

littleboobs commented on code/game/objects/items/weapons/storage/boxes.dm:
```diff
@@ -1497,6 +1497,18 @@
 	new /obj/item/ammo_box/magazine/enforcer(src)
 	new /obj/item/clothing/accessory/holster(src)
 
+/obj/item/storage/box/specter_kit
+	name = "Specter kit"
+	desc = "Коробка, содержащая пистолет \"Спектр\", кобуру и 2 обоймы парализующих патронов."
+	icon_state = "box_specter"
```
рунеймсы

PiroMage replied:
Не понял, что не так? Рунеймс, как и надо, нет?

littleboobs reviewed (changes requested):
надо доработать

Den109G commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -298,6 +298,22 @@
+	icon_state = "specstamina"
+	origin_tech = "combat=2"
+	ammo_type = /obj/item/ammo_casing/specter/disable
+	max_ammo = 30
```
А почему это не прописывать у родительского `
/obj/item/ammo_box/specter` ? Общие параметры, количество, техи и т.д.

PiroMage replied:
Хз, а что именно надо?

Den109G replied:
Прописать 
```dm
/obj/item/ammo_box/specter
   max_ammo = 30
   origin_tech = "combat=2"
```

Den109G commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -675,6 +675,20 @@
+	icon_state = "specmaglaser"
+	ammo_type = /obj/item/ammo_casing/laser
+	origin_tech = "combat=3"
+	caliber = "laser"
```
Уверен что тот калибр? У спектра такого не наблюдаю.

PiroMage replied:
Ну да, в игре все значения перенялись, на видео в дисе, которые кидал, видно.

PiroMage replied:
![image](https://github.com/user-attachments/assets/fdb67364-ada8-4c96-a94e-a4ca5bbc3389)

Den109G replied:
Этого нет на видео/или не акцентировано внимание. В описании ПРа про него ничего не написано.
По тому что я знаю - это не залезет в пистолет.
Не понимаю зачем ты скинул скрин - они ничего не подтверждают и не опровергают. Лазеры отдельно от гильз, магазинов и пушек созданы.

PiroMage committed:
Правки

pilygun commented:
Сплит опять забайтил очередного бедолагу кодера на добавление дизбалансой хуйни в билд.
А если серьезно, закинь спектр с летальными патронами на миксед и ренж тандердром)))

littleboobs commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -675,6 +675,20 @@
 
+/obj/item/ammo_box/magazine/specmaglaser
+	name = "small encased laser projector magazine"
+	desc = "Вмещает экспериментальные гильзы от лазерных боеприпасов."
```
Почему экспериментальные то, если везде просто лазерные патроны?

PiroMage replied:
Ну лазер значит что патроны лазерные, описание уже IC, что новый тип боеприпаса для лазерного оружия. Все окей.

littleboobs commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -30,6 +30,62 @@
+	INSTRUMENTAL = "Спектром",
+
+	PREPOSITIONAL = "Спектре"
+	)
```
Иксди. Имелось в виду табуляция не на строчке выше, а просто перед падежом по табуляции, посмотри в коде любой рунеймс как сделан. Код стайл.

PiroMage replied:
Дак там и так вроде табуляция, нет? Код вроде не ругался на прошлый вариант. Ну щас попробую еще раз переделать....

PiroMage replied:
А вообще нужны эти рунеймсы с падежами? Мб просто убрать их. Наверное лучше будет вообще без склонений. Не помню, чтобы где-то оно вообще требовалось для оружия.

dageavtobusnick replied:
> А вообще нужны эти рунеймсы с падежами? Мб просто убрать их. Наверное лучше будет вообще без склонений. Не помню, чтобы где-то оно вообще требовалось для оружия.

Нет, не лучше. Оно делается на будущее. Плюсом оружие это не только оружие, оно и как объект может использоваться в текстах. Рунеймсы прописывай, там где сказано их прописать.

PiroMage committed:
Убраны падежи

Мне кажется лучше вообще без них, ибо всегда будут говорить одно "Спектр", думаю склонять и не нужно.

PiroMage committed:
Последняя доработка

Собственно рунеймсы на все вещи пистолета.
От меня больше ничего не будет, да.

dageavtobusnick commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -724,3 +738,39 @@
+
+/obj/item/ammo_box/magazine/specter/examine(mob/user)
+	. = ..()
+	if(get_dist(user, src) <= 2)
```
Инвертируй условие и сделай ранний ретурн.

dageavtobusnick commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -675,6 +675,20 @@
 	icon_state = "lmag-[CEILING(ammo_count(), 5)]"
 
 
+/obj/item/ammo_box/magazine/specmaglaser
```
А что это у тебя за магазин вообще? У тебя ниже есть лазерный и не лазерный, а так же коробки, а это что?

dageavtobusnick reviewed (changes requested):
Так же обрати внимание на замечания Дена и исправь то, на что он указал

PiroMage committed:
Добивочная

1) Убран несуществующий тестовый магазин.
2) Общие переменные коробок патронов вынесены под одну базу.
3) Вернул склонения.

PiroMage committed:
Рунеймсы и код

dageavtobusnick commented on code/modules/research/designs/autolathe_designs.dm:
```diff
@@ -866,23 +866,23 @@
 
 /datum/design/specter/laser
-	name = "Коробка Патронов (Спектр Лазер)"
+	name = "коробка патронов (спектр лазер)"
```
Только первую заглавную нельзя. Остальные можно

dageavtobusnick replied:
Первую букву мы оставляем в нижнем регистре, чтобы мы могли в нужный момент сделать её заглавной, при сборке текстов. Остальные буквы статичны и могут быть в верхнем регистре, если надо.

PiroMage committed:
Еще доработки

1) Допольнительные табуляции
2) Изменены заглавные буквы в первых словах неймов и рунеймов
3) Добавлены рунеймы магазинам и коробкам патронов

PiroMage committed:
Update boxes.dm

littleboobs commented on code/game/machinery/vending.dm:
```diff
@@ -2194,6 +2194,7 @@
 		var/static/list/available_kits = list(
 			"Dominator Kit" = /obj/item/storage/box/dominator_kit,
 			"Enforcer Kit" = /obj/item/storage/box/enforcer_kit,
+			"Набор Спектр" = /obj/item/storage/box/specter_kit,
```
```suggestion
			"Dominator kit" = /obj/item/storage/box/dominator_kit,
			"Enforcer kit" = /obj/item/storage/box/enforcer_kit,
			"Specter kit" = /obj/item/storage/box/specter_kit,
```

littleboobs commented on code/game/objects/items/weapons/storage/boxes.dm:
```diff
@@ -1497,6 +1497,25 @@
 	new /obj/item/clothing/accessory/holster(src)
 
+/obj/item/storage/box/specter_kit
+	name = "набор Спектр"
```
```suggestion
	name = "Specter kit"
```

PiroMage commented:
Ну вроде все сделал...

littleboobs commented on code/modules/projectiles/ammunition/ammo_casings.dm:
```diff
@@ -374,6 +374,24 @@
 	muzzle_flash_color = LIGHT_COLOR_DARKRED
 
+/obj/item/ammo_casing/specter/laser
+	desc = "Гильза от пистолета Спектр."
```
```suggestion
	desc = "Лазерный патрон для пистолета \"Спектр\"."
```

littleboobs commented on code/modules/projectiles/ammunition/ammo_casings.dm:
```diff
@@ -374,6 +374,24 @@
+	muzzle_flash_color = LIGHT_COLOR_DARKRED
+
+/obj/item/ammo_casing/specter/disable
+	desc = "Гильза от пистолета Спектр."
```
```suggestion
	desc = "Парализующий патрон для пистолета \"Спектр\"."
```

littleboobs commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -298,6 +298,39 @@
+	max_ammo = 30
+
+/obj/item/ammo_box/specter/laser
+	name = "коробка патронов (спектр лазер)"
```
```suggestion
	name = "ammo box (Specter laser)"
```

littleboobs commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -298,6 +298,39 @@
+
+
+/obj/item/ammo_box/specter/disabler
+	name = "коробка патронов (спектр дизейблер)"
```
```suggestion
	name = "ammo box (Specter disabler)"
```

littleboobs commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -724,3 +724,56 @@
+/obj/item/ammo_box/magazine/specter
+	name = "магазин спектр (дизейблер)"
+	desc = "Магазин пистолета \"Спектр\", предназначенный для парализующих патронов."
+	ru_names = list(
+		NOMINATIVE = "магазин Спектр (дизейблер)",
+		GENITIVE = "магазин Спектр (дизейблер)",
+		DATIVE = "магазин Спектр (дизейблер)",
+		ACCUSATIVE = "магазин Спектр (дизейблер)",
+		INSTRUMENTAL = "магазин Спектр (дизейблер)",
+		PREPOSITIONAL = "магазин Спектр (дизейблер)"
+	)
```
не дизейблер, а парализующий

littleboobs commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -724,3 +724,56 @@
 	max_ammo = 14
+
+/obj/item/ammo_box/magazine/specter
+	name = "магазин спектр (дизейблер)"
```
```suggestion
	name = "magazine Specter (disabler)"
```

littleboobs commented on code/modules/projectiles/ammunition/boxes.dm:
```diff
@@ -298,6 +298,39 @@
+/obj/item/ammo_box/specter/disabler
+	name = "коробка патронов (спектр дизейблер)"
+	desc = "Коробка, содержащая 30 парализующих патронов для пистолета \"Спектр\"."
+	ru_names = list(
+		NOMINATIVE = "коробка патронов (Спектр дизейблер)",
+		GENITIVE = "коробка патронов (Спектр дизейблер)",
+		DATIVE = "коробке патронов (Спектр дизейблер)",
+		ACCUSATIVE = "коробку патронов (Спектр дизейблер)",
+		INSTRUMENTAL = "коробкой патронов (Спектр дизейблер)",
+		PREPOSITIONAL = "коробке патронов (Спектр дизейблер)"
+	)
```
\"Спектр\" парализующие

littleboobs commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -724,3 +724,56 @@
+	return FALSE
+
+/obj/item/ammo_box/magazine/specter/laser
+	name = "магазин спектр (лазер)"
```
```suggestion
	name = "magazine Specter (laser)"
```

littleboobs commented on code/modules/projectiles/ammunition/magazines.dm:
```diff
@@ -724,3 +724,56 @@
+/obj/item/ammo_box/magazine/specter/laser
+	name = "магазин спектр (лазер)"
+	desc = "Магазин пистолета \"Спектр\", предназначенный для лазерных патронов."
+	ru_names = list(
+		NOMINATIVE = "магазин Спектр (лазер)",
+		GENITIVE = "магазин Спектр (лазер)",
+		DATIVE = "магазин Спектр (лазер)",
+		ACCUSATIVE = "магазин Спектр (лазер)",
+		INSTRUMENTAL = "магазин Спектр (лазер)",
+		PREPOSITIONAL = "магазин Спектр (лазер)"
+	)
```
лазерный

littleboobs commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -30,6 +30,56 @@
+
+//Specter//
+/obj/item/gun/projectile/automatic/pistol/specter
+	name = "спектр"
```
```suggestion
	name = "Specter"
```

littleboobs commented on code/modules/projectiles/guns/projectile/pistol.dm:
```diff
@@ -30,6 +30,56 @@
+/obj/item/gun/projectile/automatic/pistol/specter
+	name = "спектр"
+	desc = "Современный пистолет \"Спектр\", модернизирован для возможности стрельбы лазерными патронами. Поставляется только силовым структурам Нанотрейзен."
+	ru_names = list(
+	NOMINATIVE = "спектр",
+	GENITIVE = "спектра",
+	DATIVE = "спектру",
+	ACCUSATIVE = "спектр",
+	INSTRUMENTAL = "спектром",
+	PREPOSITIONAL = "спектре"
+	)
```
Табуляция и с заглавной буквы везде. Это имя, правила русского языка.

littleboobs commented on code/modules/research/designs/autolathe_designs.dm:
```diff
@@ -865,6 +865,39 @@
 	build_path = /obj/item/ammo_box/c9mmte
 	category = list("hacked", "Security")
 
+/datum/design/specter/disable
+	name = "коробка патронов (Спектр дизейблер)"
+	id = "specter_disable"
+	build_type = AUTOLATHE
+	materials = list(MAT_METAL = 25000)
+	build_path = /obj/item/ammo_box/specter/disabler
+	category = list("initial", "Security")
+
+/datum/design/specter/laser
+	name = "коробка патронов (Спектр лазер)"
+	id = "specter_laser"
+	build_type = AUTOLATHE
+	materials = list(MAT_METAL = 35000)
+	build_path = /obj/item/ammo_box/specter/laser
+	category = list("hacked", "Security")
```
нейм на английском, анлогично тому, что указал раннее в саггестах

littleboobs commented on code/modules/research/designs/autolathe_designs.dm:
```diff
@@ -865,6 +865,39 @@
+	build_path = /obj/item/ammo_box/specter/laser
+	category = list("hacked", "Security")
+
+/datum/design/spectermag_disabler
+	name = "магазин Спектр (дизейблер)"
+	id = "spectermag_disabler"
+	build_type = AUTOLATHE
+	materials = list(MAT_METAL = 8000)
+	build_path = /obj/item/ammo_box/magazine/specter
+	category = list("initial", "Security")
+
+//hacked autolathe recipes
+/datum/design/spectermag_laser
+	name = "магазин Спектр (лазер)"
+	id = "spectermag_laser"
+	build_type = AUTOLATHE
+	materials = list(MAT_METAL = 10000)
+	build_path = /obj/item/ammo_box/magazine/specter/laser
+	category = list("hacked", "Security")
```
нейм на английском аналогично

littleboobs reviewed (changes requested):
Утомляющий ПР. Вся локализация была готова в старом ПРе буквально. С чего-то перевёл все неймы зачем-то. Почти всё исправил за тебя. Исправь остальное, указал где.
Брух.

PiroMage committed:
Дополнения

1) Доп. табуляции в рунеймсах.
2) Нейм на англ.
3) Рунейм с большой буквы.

PiroMage committed:
Merge branch 'Specter-pistor' of https://github.com/PiroMage/Paradise into Specter-pistor

PiroMage committed:
Update boxes.dm

Коробка патронов (Спектр лазер) => Коробка патронов (Спектр лазерные)

littleboobs reviewed (approved):
Всё поправил, красава!
Прожимай ресолв на комментарии, которые решаешь.
